src/lib/optimizer/strategy/constant_calculation_rule.cpp

```diff
@@ -47,54 +62,107 @@ bool ConstantCalculationRule::apply_to(const std::shared_ptr<AbstractLQPNode>& n
-    if (value != std::nullopt &&
-        _replace_expression_in_parents(node, node->output_column_references()[column_id], *value)) {
-      // If we successfully replaced the occurrences of the expression in the parent tree,
-      // we can remove the column here.
-      _remove_column_from_projection(projection_node, column_id);
+    if (value != std::nullopt) {
+      column_reference_to_value_map.emplace(projection_node->output_column_references()[column_id], *value);
     }
   }
-
-  return _apply_to_children(node);
 }
 
-bool ConstantCalculationRule::_replace_expression_in_parents(const std::shared_ptr<AbstractLQPNode>& node,
-                                                             const LQPColumnReference& expression_column,
-                                                             const AllTypeVariant& value) {
-  auto parent_tree_changed = false;
-  for (auto parent : node->parents()) {
-    if (parent->type() != LQPNodeType::Predicate) {
-      parent_tree_changed |= _replace_expression_in_parents(parent, expression_column, value);
-      continue;
-    }
+bool ConstantCalculationRule::_replace_column_references_in_tree(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    const std::map<LQPColumnReference, AllTypeVariant>& column_reference_to_value_map) {
+  auto tree_changed = false;
 
-    auto predicate_node = std::static_pointer_cast<PredicateNode>(parent);
+  if (node->left_child()) {
+    tree_changed |= _replace_column_references_in_tree(node->left_child(), column_reference_to_value_map);
+  }
+  if (node->right_child()) {
+    tree_changed |= _replace_column_references_in_tree(node->right_child(), column_reference_to_value_map);
+  }
+
+  if (node->type() == LQPNodeType::Predicate) {
+    auto predicate_node = std::static_pointer_cast<PredicateNode>(node);
 
     // We look for a PredicateNode which has an LQPColumnReference as value,
     // referring to the column which contains the Expression we resolved before.
-    if (is_lqp_column_reference(predicate_node->value()) &&
-        boost::get<LQPColumnReference>(predicate_node->value()) == expression_column) {
-      // We replace the LQPColumnReference with the actual result of the Expression it was referring to.
-      auto new_predicate_node = std::make_shared<PredicateNode>(predicate_node->column_reference(),
-                                                                predicate_node->predicate_condition(), value);
-      predicate_node->replace_with(new_predicate_node);
-      parent_tree_changed = true;
+    if (is_lqp_column_reference(predicate_node->value())) {
+      const auto column_reference = boost::get<LQPColumnReference>(predicate_node->value());
+      if (column_reference_to_value_map.find(column_reference) != column_reference_to_value_map.end()) {
+        // We replace the LQPColumnReference with the actual result of the Expression it was referring to.
+        auto new_predicate_node =
```

Comment:
`predicate_node` might be a BETWEEN predicate, this needs to be handled as well. (Probably by just passing value2 of the original predicate node into the constructor of the new predicate_node)